serverless: read a configuration file while running as a Lambda Extension.

remeh opened the pull request:
### What does this PR do?

The Datadog Lambda Extension will try to read the configuration
file in `/var/task/datadog.yaml`.

### Motivation

It will help us use some nice features such as a the logs scrubbing.

### Additional Notes

I've documented its usage in the matching `README.md`.
I've also removed the expvar HTTP server which is not used by the Serverless Agent.

### Describe your test plan

N/A

remeh committed:
serverless: do not start the HTTP server for expvar.

It is not used while running as a Serverless Agent.

remeh committed:
serverless: read a configuration file while running as a Lambda Extension.

The Datadog Lambda Extension will try to read the configuration
file in `/var/task/datadog.yaml`.
It will let us use some nice features such as a the logs scrubbing.

remeh committed:
serverless: document the logs filtering and scrubbing support.

DarcyRaynerDD commented on cmd/serverless/main.go:
```diff
@@ -231,10 +221,11 @@ func runAgent(ctx context.Context, stopCh chan struct{}) (err error) {
-		log.Warn("A configuration file has been found, which should not happen in this mode.")
+	// we will look for the configuration file in the /var/task folder,
+	// where AWS Lambda should make it available.
+	config.Datadog.SetConfigFile("/var/task/datadog.yaml")
```
We could reuse the environment variable: 'datadogConfigPath'. Also, looking at the code I believe there may be a bug with how the trace agent config is loaded. The value of 'datadogConfigPath' should be "/var/task/datadog.yaml"

remeh replied:
I've changed the value of the const and I'm now using it! This should address both concerns.

remeh committed:
serverless: use the const to look for the configuration file instead of hardcoded value.

I've changed the const to look for the file into `/var/task` where it should belongs
on AWS Lambda. Note that the trace agent is also using this const.

nhinsch commented:
Looking at the [example configuration file](https://github.com/DataDog/datadog-agent/blob/master/pkg/config/config_template.yaml), I see that there are a *lot* of possible options. Presumably not all of these apply in the Serverless context. Should we document which options are supported? What is the worst that can happen if an unsupported option is added?

nhinsch commented on cmd/serverless/main.go:
```diff
@@ -231,10 +224,9 @@ func runAgent(ctx context.Context, stopCh chan struct{}) (err error) {
 	// read configuration from the environment vars
```
I would update this comment

remeh replied:
Good catch. Updating that.

remeh committed:
serverless: update comment indicating where is read the configuration

remeh commented:
@nhinsch that's a very good question: before sending that PR I took a quick look at that file as well. I think that for most of these parameters, not much unintended would happen or at least nothing surprising, ofc if some of them are wrongly used the extension will misbehave and a lot of them makes no sense in the context of the extension.

My opinion is that we should document only the ones that make sense but I don't feel like we need to implement a whitelist, but it's really not a strong opinion. I like the idea that we may be able to leverage some of these existing fields in the future without having to release a new version.

hush-hush commented on cmd/serverless/README.md:
````diff
@@ -13,7 +13,52 @@ GOOS=linux go build -tags serverless -o datadog-agent
+Datadog Agent. All you have to do is to set `processing_rules` in the `logs_config`
+field of your configuration.
+
+For example, if you want to redact some parts of the logs collected by the Datadog Lambda
+Extension / Serverless Agent, you can use this syntax:
+
+```
+logs_config:
+    processing_rules:
+        # replace sequences in the logs
+        - type: mask_sequences
+          name: mask_user_email
+          replace_placeholder: "MASKED_EMAIL@datadoghq.com"
+          pattern: \w+@datadoghq.com
+        - type: mask_sequences
+          name: mask_credit_cards
+          replace_placeholder: "[masked_credit_card]"
+          pattern: (?:4[0-9]{12}(?:[0-9]{3})?|[25][1-7][0-9]{14}|6(?:011|5[0-9][0-9])[0-9]{12}|3[47][0-9]{13}|3(?:0[0-5]|[68][0-9])[0-9]{11}|(?:2131|1800|35\d{3})\d{11})
+```
+
+Another example, if you want to exclude some logs line containing a pattern:
+
+```
+logs_config:
+    processing_rules:
+    - type: exclude_at_match
+      name: exclude_datadoghq_users
+      # regexp can be anything
+      pattern: \w+@datadoghq.com
+```
````
Do we want duplicate example from the doc ? In case of a documentation update this might not be edited. I think the idea is to have one place of truth with the documentation website.

remeh replied:
I also had mixed feeling concerning shipping samples here copied from the doc. I'll remove them.

hush-hush commented on cmd/serverless/main.go:
```diff
@@ -85,7 +84,10 @@ where they can be graphed on dashboards. The Datadog Serverless Agent implements
 	logsLogsTypeSubscribed = "DD_LOGS_CONFIG_LAMBDA_LOGS_TYPE"
 
-	datadogConfigPath        = "datadog.yaml"
+	// AWS Lambda is writing the Lambda function files in /var/task, we want the
```
Is this definitive from AWS ? Could we let user modify this through an env var (without documenting it for now). This could be useful if one day AWS chose to change this to avoid a force upgrade of the agent.

remeh replied:
I think it's fairly definitive indeed since it is how it works and how it is documented on AWS Lambda environments, it was already the case even before the appearance of the extension stuff. We would probably be noticed in time if such a big change would occur (+ releasing a new layer for that is shorter than a full agent release for now).

hush-hush commented on cmd/serverless/main.go:
```diff
@@ -228,13 +221,13 @@ func runAgent(ctx context.Context, stopCh chan struct{}) (err error) {
-	if _, confErr := config.Load(); confErr == nil {
-		log.Warn("A configuration file has been found, which should not happen in this mode.")
+	config.Datadog.SetConfigFile(datadogConfigPath)
+	if _, confErr := config.LoadWithoutSecret(); confErr == nil {
```
Could people add their secret executable to the lambda root and be able to use the secret function ? This might be for another PR.

remeh replied:
Discussed offline but for now, we won't use the `secrets` feature in the Lambda Extension. It should not be a big deal to enable it but it would be for another PR 👍

hush-hush reviewed (approved):
Added a few questions, looks good otherwise.

remeh committed:
serverless: remove logs scrubbing configuration samples.

We don't want the repository doc to be out of sync with public documentation.

remeh committed:
Merge remote-tracking branch 'origin/master' into remeh/serverless-logs-scrubs
